crates/pallets/starknet/src/lib.rs

```diff
@@ -775,25 +777,36 @@ pub mod pallet {
 
             // Validate the user transactions
             if let UserAndL1HandlerTransaction::User(transaction) = transaction {
-                match transaction {
-                    UserTransaction::Declare(tx, contract_class) => tx
-                        .try_into_executable::<T::SystemHash>(chain_id, contract_class, false)
-                        .map_err(|_| InvalidTransaction::BadProof)?
-                        .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
-                    // There is no way to validate it before the account is actuallly deployed
-                    UserTransaction::DeployAccount(_) => Ok(None),
-                    UserTransaction::Invoke(tx) => tx.into_executable::<T::SystemHash>(chain_id, false).validate_tx(
-                        &mut state,
-                        &block_context,
-                        &mut execution_resources,
-                        &mut initial_gas,
-                        false,
-                    ),
-                }
-                .map_err(|e| {
-                    log::error!("failed to validate tx: {}", e);
-                    InvalidTransaction::BadProof
-                })?;
+                let validate_result =
+                    match transaction {
+                        UserTransaction::Declare(tx, contract_class) => tx
+                            .try_into_executable::<T::SystemHash>(chain_id, contract_class, false)
+                            .map_err(|_| InvalidTransaction::BadProof)?
+                            .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+                        // There is no way to validate it before the account is actually deployed
+                        UserTransaction::DeployAccount(_) => Ok(None),
+                        UserTransaction::Invoke(tx) => tx
+                            .into_executable::<T::SystemHash>(chain_id, false)
+                            .validate_tx(&mut state, &block_context, &mut execution_resources, &mut initial_gas, false),
+                    };
+                match validate_result {
+                    Ok(_) => {}
+                    Err(TransactionExecutionError::ValidateTransactionError(
+                        EntryPointExecutionError::PreExecutionError(PreExecutionError::UninitializedStorageAddress(
+                            contract_address,
+                        )),
+                    )) => {
+                        // If the the txn is failing because the account isn't deployed yet
+                        // then don't throw an error and add a `require` tag for transaction
+                        // of nonce 0 to be executed first.
+                        if contract_address.0.0 != sender_address.into()
+                            || transaction_nonce != Some(Felt252Wrapper::ONE)
+                        {
+                            Err(InvalidTransaction::BadProof)?
```

Comment:
Is there any point to differentiate between the error when the nonce isn't 0 or 1, and other cases? Like below the error says the same, `BadProof`. Could be more verbose, wdyt?

Reply:
There is no error thrown when the nonce is 1 and the failure reason is `UninitializedStorageAddress`. All other cases its `BadProof`